File: app-foretime/src/main/java/com/smallraw/foretime/app/repository/db/dao/MemorialDao.java
package com.smallraw.foretime.app.repository.db.dao;

import android.arch.persistence.db.SupportSQLiteQuery;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.RawQuery;
import android.arch.persistence.room.Update;

import com.smallraw.foretime.app.repository.db.entity.MemorialDO;

import java.util.List;

@Dao
public interface MemorialDao {
    @RawQuery
    List<MemorialDO> select(SupportSQLiteQuery sql);

    @Query("SELECT COUNT(*) FROM memorial")
    int count();

    @Insert
    long insert(MemorialDO memorialDO);

    @Insert
    List<Long> insertAll(MemorialDO[] memorialDO);

    @Query("SELECT * FROM memorial")
    List<MemorialDO> selectAll();

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0")
    List<MemorialDO> selectActive();

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0 AND type = :display ORDER BY " +
            ":order DESC")
    List<MemorialDO> selectOptionActive(int display, String order);

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0 ORDER BY :order DESC")
    List<MemorialDO> selectOptionActive(String order);

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0 ORDER BY targetTime DESC")
    List<MemorialDO> selectOptionActive();

    @Query("SELECT * FROM memorial WHERE id = :id")
    MemorialDO selectById(long id);

    @Update
    int update(MemorialDO memorialDO);

    @Update
    int update(List<MemorialDO> memorialDO);

    @Query("DELETE FROM memorial WHERE id = :id")
    int deleteById(long id);

    @Query("DELETE FROM memorial")
    int deleteAll();

    @Delete
    int deletes(List<MemorialDO> memorialDO);
}

File: app-foretime/src/main/java/com/smallraw/foretime/app/repository/db/dao/MemorialDao.kt
package com.smallraw.foretime.app.repository.db.dao

import android.arch.persistence.db.SupportSQLiteQuery
import android.arch.persistence.room.*
import com.smallraw.foretime.app.repository.db.entity.MemorialDO

@Dao
interface MemorialDao {
    @RawQuery
    fun select(sql: SupportSQLiteQuery): List<MemorialDO>

    @Query("SELECT COUNT(*) FROM memorial")
    fun count(): Int

    @Insert
    fun insert(memorialDO: MemorialDO): Long

    @Insert
    fun insertAll(memorialDO: Array<MemorialDO>): List<Long>

    @Query("SELECT * FROM memorial")
    fun selectAll(): List<MemorialDO>

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0")
    fun selectActive(): List<MemorialDO>

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0 AND type = :display ORDER BY " + ":order DESC")
    fun selectOptionActive(display: Int, order: String): List<MemorialDO>

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0 ORDER BY :order DESC")
    fun selectOptionActive(order: String): List<MemorialDO>

    @Query("SELECT * FROM memorial WHERE strike = 0 AND archive = 0 ORDER BY targetTime DESC")
    fun selectOptionActive(): List<MemorialDO>

    @Query("SELECT * FROM memorial WHERE id = :id")
    fun selectById(id: Long): MemorialDO

    @Update
    fun update(memorialDO: MemorialDO): Int

    @Update
    fun update(memorialDO: List<MemorialDO>): Int

    @Query("DELETE FROM memorial WHERE id = :id")
    fun deleteById(id: Long): Int

    @Query("DELETE FROM memorial")
    fun deleteAll(): Int

    @Delete
    fun deletes(memorialDO: List<MemorialDO>): Int
}